### PR TITLE
UI Notifications for errors

### PR DESCRIPTION
**Pull request by DivyanshiSingh**
Fixes: #13 and #101 
1) Created a reuseable notification component in patternfly-4 to display the error messages.

### CODE REVIEW

File: src/errors/HttpApiCallError.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export default class HttpApiCallError extends Error {
+  constructor(error) {
+    super(error);
+    this.message = error.message;
```
Comment (jennydaman) [inline src/errors/HttpApiCallError.js:4]: Could use object spread syntax here (optional).

File: src/components/Plugins/Plugins.js [hunk 2]
```diff
@@ -52,7 +55,8 @@ export class Plugins extends Component {
 
   componentDidMount() {
     this.fetchPlugins().catch((err) => {
-      console.error(err);
+      console.log("here")
```
Comment (jennydaman) [inline src/components/Plugins/Plugins.js:58]: delete `console.log`

**Review by jennydaman — COMMENTED** [review]
Only nitpicks, see review. 

Preview

File: src/components/Dashboard/Dashboard.js [hunk 3]
```diff
@@ -61,6 +69,7 @@ class Dashboard extends Component {
         this.fetchPlugins();
       });
     } catch (e) {
+      this.showNotifications(new HttpApiCallError(e));
       return e;
```
[the remark this answers is not in the document]
Comment (DivyanshiSingh) [reply src/components/Dashboard/Dashboard.js:73]: We can either use custom messages as I was using earlier or use the custom response from the backend, but as you said that the backend response is not structured, I think it's best to use the custom messages.

File: src/components/Plugins/Plugins.js (hunk 2 above)
Comment (DivyanshiSingh) [reply src/components/Plugins/Plugins.js]: console deleted.

File: src/components/Dashboard/Dashboard.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jennydaman) [reply src/components/Dashboard/Dashboard.js:73]: We can detect some custom messages simply based on the response (404, wrong content type, timeout, ...)
However that is besides the point. What you added is okay. I was making a comment on the line below, how the code reads

```javascript
catch (e) {
  something();
  return e;
}
```

In general, we shouldn't return anything in the case of an exception, we should either throw it again or give up and do nothing. If we return an exception, we are making a bad situation worse.

However, I took a moment to scroll up and actually read the existing code. It's async spaghetti. The `try` is only wrapping these lines of code:

https://github.com/FNNDSC/ChRIS_store_ui/pull/144/files#diff-e82a9b05e2166cb4373028e8db04cb652be90796eb324ad21dbbf2f5ef9588a9L30-L46

The actual request and errors regarding the request occur during a `Promise`.

I think what was intended here was,

```javascript
client.getPlugin(pluginId)
  .then(plugin => plugin.delete())
  .then(() => this.fetchPlugins())
  .catch(e => this.showNotifications(new HttpApiCallError(e)));
```

Can be cleaner using `async`/`await`. Do you think you could try rewriting this using `async`/`await`?

**Review by jennydaman — CHANGES_REQUESTED** [review]

File: src/components/Dashboard/Dashboard.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (DivyanshiSingh) [reply src/components/Dashboard/Dashboard.js:73]: thanks for pointing this out. I will remove the return statement and rewrite using async/await

[the remark this answers is not in the document]
Comment (DivyanshiSingh) [reply src/components/Dashboard/Dashboard.js]: changes done please review

File: src/components/Dashboard/Dashboard.js [hunk 4]
```diff
@@ -49,20 +57,21 @@ class Dashboard extends Component {
...
     const storeURL = process.env.REACT_APP_STORE_URL;
     const auth = { token: store.get('authToken') };
     const client = new Client(storeURL, auth);
 
     let response;
     try {
-      response = client.getPlugin(pluginId).then(plugin => plugin.delete());
-      response.then(() => {
+      response = await client.getPlugin(pluginId);
+      response.delete();
```
Comment (jennydaman) [inline src/components/Dashboard/Dashboard.js:69]: Let's  `await response.delete()` as well.

**Review by jennydaman — CHANGES_REQUESTED** [review]
Sorry for the late responses @DivyanshiSingh. This PR is almost ready.

File: src/components/Dashboard/Dashboard.js (hunk 4 above)
Comment (manas2297) [reply src/components/Dashboard/Dashboard.js]: done

File: src/components/Dashboard/Dashboard.js [hunk 5]
```diff
@@ -49,20 +57,21 @@ class Dashboard extends Component {
...
     let response;
     try {
-      response = client.getPlugin(pluginId).then(plugin => plugin.delete());
-      response.then(() => {
+      response = await client.getPlugin(pluginId);
+      await response.delete();
+      if(response.data) {
         this.fetchPlugins();
-      });
+      }
```
Comment (jennydaman) [inline src/components/Dashboard/Dashboard.js:72]: else, maybe throw an error to trigger UI notification?

File: src/components/Plugin/Plugin.js [hunk 6]
```diff
@@ -41,12 +44,15 @@ export class Plugin extends Component {
...
     let pluginData;
-    if (!this.state.pluginData) {
-      pluginData = await this.fetchPluginData();
-    } else {
-      ({ pluginData } = this.state);
+    try{
+      if (!this.state.pluginData) {
+        pluginData = await this.fetchPluginData();
+      } else {
+        ({ pluginData } = this.state);
```
Comment (jennydaman) [inline src/components/Plugin/Plugin.js:51]: Would you remove these parenthesis?

File: src/components/Plugin/Plugin.js [hunk 7]
```diff
@@ -114,18 +126,20 @@ export class Plugin extends Component {
...
+    try{
+      if (auth) {
+        const client = new Client(storeURL, auth);
+        const response = await client.getPluginStars({ plugin_name: pluginName });
+        stars = response.data;
+      }
+  
+      if (stars.length > 0) {
+        this.setState({ star: stars[0] });
+      }
```
Comment (jennydaman) [inline src/components/Plugin/Plugin.js:138]: else... throw

Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:138]: Can you specify more about what to throw here?

Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js:138]: done

File: src/components/Plugin/Plugin.js (hunk 6 above)
Comment (DivyanshiSingh) [reply src/components/Plugin/Plugin.js]: done

File: src/components/Dashboard/Dashboard.js (hunk 5 above)
Comment (DivyanshiSingh) [reply src/components/Dashboard/Dashboard.js:72]: done

**Review by jennydaman — APPROVED** [review]
I myself would be more specific about errors and exception handling (either "unexpected response from server," or in the worst case, "unknown error"), but being more specific about error handling, though it might not be helpful at this level. What you have here greatly improves UX and is good to merge.

**Comment by DivyanshiSingh** [comment]
Hey, @jennydaman do you have some specific feature in mind that you feel is primarily important to add in pf4. I can take that up in parallel to form. I will raise the PR for the form by today.

**Comment by jennydaman** [comment]
@DivyanshiSingh it's definitely still about eliminating all trace of `patternfly-react` from the codebase, and maybe upgrading to react v17. They're chores but we must pay off the technical debt before working on the features.

**Pull request merged**